rep-2016.rst

```diff
@@ -0,0 +1,735 @@
+
+Therefore the implicit intermediary output looks like:
+
+.. code::
+
+    ServiceType_Request request_message
+    ServiceType_Response response_message
+    ServiceType_Event event_message
+
+And Actions contain
+
+- a Goal message as defined in input
+- a Result message as defined in input
+- a Feedback message as defined in input
+- a SetGoal service, derived
+- a GetResult service, derived
+- a FeedbackMessage, derived
+
+Therefore their implicit intermediary output looks like:
+
+.. code::
+
+    ActionType_Goal goal
+    ActionType_Result result
+    ActionType_Feedback feedback
+    ActionType_SendGoalService send_goal_service
+    ActionType_GetResultService get_result_service
+    ActionType_FeedbackMessage feedback_message
+
+
+While this method is simple and therefore recommended, it is worth noting that it has the following drawback:
+A message defined explicitly with those above field types and names would have an identical description to a Service or Action, even though they are different category of interfaces.
+However, this case is fairly contrived and in practice an application will already know what interface category it is trying to interact with.
+
+Given this approach, the ``type_description_interfaces`` provides no special accommodation for Services and Actions, leaving the transform into a representable type to code generation.
+The reference implementation for this derivation is in ``rosidl_generator_type_description``\ [#td_gen]_.
+
+Versioning the ``TypeDescription`` Message Type
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Given that the type description message interface has to be generic enough to support anything described in the ROS interfaces, there will be a need to add or remove fields over time in the type description message itself.
+This should be done in such a way that the fields are tick-tocked and deprecated properly.
+It is recommended to do this by having explicitly named versions of this interface, e.g. ``TypeDescriptionV2``, ``TypeDescriptionV3`` and so on.
+The initial version is named without a version identifier.
+Revisions may be inevitable, but there should be a default preference against modifying, to keep the number of these versions to a minimum.
+
+``TypeDescription`` Self-Describing Loop Problem
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This section covers an implementation detail rather than a goal-oriented design decision, but is worth covering as it represents a significant difficulty.
+
+The ``TypeDescription`` messages are defined as just another ROS interface type, using all the same tooling.
+This means that its data structures are not available yet, during build time, to describe itself and other interface packages.
+The final decision to support this is a checked-in mirror of the generated type structs, provided in ``rosidl_runtime_c``, for use by C (used by Python) and C++ code generation.
```

Comment:
(question) as user who builds the source code with their own CI/CD pipeline, descriptions are generated during build time, including `TypeDescription` messages. these descriptions will be the same with the ones from released binary packages as long as source code is the same version. just checking my understanding is correct.

Reply:
Yes, that is correct! As long as the source code is the same version, then the `TypeDescriptions` and hashes will come out the same, they are deterministic.